src/xview/scantool/scantool.1
.\" @(#)scantool.1 1.3 90/04/04
.TH SCANTOOL 1L "2 April 1990"
.SH NAME
scantool \- a program to scan images using a Microtek MS300A scanner.
.SH SYNOPSIS
.B "scantool
[
.B -v
]
[
.B \-Wi
]
[
.B \-Wp
.I x y
]
[
.B \-WP
.I x y
]
.SH DESCRIPTION
.I Scantool
allows the user to scan in documents using a Microtek
MS300A scanner, and turn the resulting image into a Sun rasterfile.
This program uses either SunView or XView graphics. These graphics
routines are fairly trivial, and have been isolated into a single
file, so it would be relatively easy to write a set of graphics
routines for another graphics system.
.LP
When you run scantool, a window is displayed. There is a black bar
at the top with containing six headings. Hitting a mouse button when
the cursor is above one of those six headings, will display a popup
menu. These options are explained in more detail below, and also by
selecting entries from the Help popup menu.
.LP
There are also four box toggle switches. Each toggle has two settings.
There are toggles for setting the line mode,
data transfer method, serial port and baud rate.
.LP
There are three buttons. The Scan button initiates a scan. Should you
wish to cancel a scan once it has been started, then you should press 
the Canel button.
.LP
The Show button allows you to display the scanned rasterfile images in 
a separate window.
.SS "Display mode."
.LP
The scanner can handle input from the image sensor in
two modes: Line Art mode and Halftone mode. Line Art
mode produces an all black-and-white image, while
Halftone mode results in an image that appears to contain
varying shades of gray.
.LP
In Line Art mode, the amount of light falling on the
sensor is compared against a "threshold" determined by
the Brightness and Contrast controls. Whenever it falls
below the threshold, a black dot is sent to the Sun; when
it exceeds the threshold a white dot is sent. This mode
is normally used for material in which everything is
either black or white.
.LP
Halftone mode is used to detect and reproduce finer
differences in shading. In this mode the dots sent to
the Sun are organised into squares, or "grains", of
uniform size. By varying the ratio of black and white
dots in the grains, different degrees of shading, or
"gray levels", are achieved.
.SS "Data transfer."
.LP
In order to reduce the amount of time needed to
transmit a scanned image to the computer, the
MS-300A can compress image data using a one-
dimensional coding scheme based on CCITT
Recommendation T.4 (modified Huffman code). By default the scanner
will send compressed data to the computer.
.SS " Serial connection."
.LP
The scanner connects to the computer using an RS232 connection.
This is currently setup to allow you to select between the RS232
A and B ports.  The baud rate can also be toggled between 9600
and 1920 baud.
.SS "Brightness."
.LP
There are 14 Brightness levels to choose from. The
effect of this control depends on whether the image
is scanned in Line Art or Halftone mode.
.LP
In Line Art mode, the higher the Brightness setting
the darker a spot must be in order for the scanner to
consider it "black". Lower settings can therefore be
used to get a sharper image if the original is too
light or blurry, while higher settings can be used to
prevent slightly shaded areas from coming out black
in the scanned image.
.LP
In Halftone mode, the Brightness setting affects
shading in all areas of the scanned image, both dark
and light. The higher the value selected, the lighter
the overall image; the lower the setting, the darker image.
.SS "Contrast."
.LP
The Contrast control also has 14 different settings.
The effect of this control in Line Art mode is identical
to that of the Brightness control. In Halftone mode, a
higher Contrast setting causes all but the darkest areas
of the original document to come out lighter in the
scanned image. Too high a setting can result in "glare
spots" of pure white; too low a setting can make the
image almost uniformly gray.
.SS "Grain."
.LP
All the grains in any one halftone image are the same
size (see the Mode help). It is desirable, however, to
use different grain sizes for different kinds of images.
The Grain control makes it possible to do this. Six
different grain sizes are available: 2, 3, 4, 5, 6, and
8 dots square.
.LP
When smaller grain sizes (i.e. higher settings) are
used, fine details in the original document are captured
more clearly, but the smaller number of dot positions in
the grains limits the scanner's "palette" of gray levels.
A Grain setting of 11, for instance produces grains that
are 2 dots high by 2 dots wide. With only 4 dot positions
to fill, only 5 distinct gray levels are possible, from
all white to all black.
.SS "Resolution."
.LP
The Resolution control determines the degree of
reduction of the scanned image relative to the original.
Since the dots that make up the scanned image cannot be
reduced in size, reduction is achieved by deleting dots
at selected intervals when transmitting the image to
the Sun, resulting in an effective resolution of less
than the scanner's maximum of 300 dots per inch.
.LP
There are 16 possible Resolution settings, providing
scaling operations from 25% to 100%.
.SS "Setting the picture name."
.LP
There are various options that can be set via the Set popup menu.
One of these is the picture name. This is the name of the file on 
the Sun that the scanned picture will be saved in. The default is 
a file called "Noname.rf".
.SS "Setting the scaning frame."
.LP
The frame is the area to be scanned. Scantool allows
you to scan an area as small as 1/8" x 1/8", or as large
as 8.5" x 11". Using the Set Scanning Frame function in
the Set menu, you can adjust the size and location of
the frame in order to scan any part of the document that
you feed into the scanner.
.SS "Initiating a scan."
.LP
When you click the Scan button, scantool checks
to see if the scanner is on line, i.e. powered up,
properly connected and ready to receive commands.
.LP
If everything checks out OK, the current settings
will be transmitted to the scanner, and the rollers
will begin to turn. Actual scanning will begin after
a document is detected by the paper sensor, which is
located in the middle of the feed slot just below
the top rollers.
.SH OPTIONS
.TP
.B \-v
Print the version number of this release of the scantool program.
.TP
.B \-Wi
Start the
.B scantool
program up in iconic form.
.TP
.BI \-Wp " x y"
Start the open window position at
.I x y
.TP 
.BI \-WP " x y"
Start the icon position at
.I x y
.SH FILES
.TP
/usr/local/lib/scantool/scantool.help
.SH AUTHOR
Rich Burridge,        Domain: richb@Aus.Sun.COM
.nf
PHONE: +61 2 413 2666   Path: {uunet,mcvax,ukc}!munnari!sunaus.oz!richb
.fi


src/NeWS
../NeWS

src/xview/scantool/scantool.c

/*  @(#)scantool.c 1.4 90/04/04
 *
 *  Main routine for scantool, which will read a scanned image
 *  from a Microtek MS-300A scanner and convert it to a Sun rasterfile.
 *
 *  Copyright (c) Rich Burridge.
 *                Sun Microsystems, Australia - All rights reserved.
 *
 *  Permission is given to distribute these sources, as long as the
 *  copyright messages are not removed, and no monies are exchanged.
 *
 *  No responsibility is taken for any errors or inaccuracies inherent
 *  either to the comments or the code of this program, but if
 *  reported to me, then an attempt will be made to fix them.
 */

#include <stdio.h>
#include <strings.h>
#include "scantool.h"

char *br_strs[] = {     /* Brightness menu item strings. */
       "1   -24% Darker",
       "2   -20% Darker",
       "3   -16% Darker",
       "4   -12% Darker",
       "5   -8%  Darker",
       "6   -4%  Darker",
       "7   No adjustment",
       "8   +4%  Lighter",
       "9   +8%  Lighter",
       "10  +12% Lighter",
       "11  +16% Lighter",
       "12  +20% Lighter",
       "13  +24% Lighter",
       "14  +28% Lighter",
       NULL
} ;

char *con_strs[] = {    /* Contrast menu item strings. */
       "1   -24% Lower",
       "2   -20% Lower",
       "3   -16% Lower",
       "4   -12% Lower",
       "5   -8%  Lower",
       "6   -4%  Lower",
       "7   No adjustment",
       "8   +4%  Higher",
       "9   +8%  Higher",
       "10  +12% Higher",
       "11  +16% Higher",
       "12  +20% Higher",
       "13  +24% Higher",
       "14  +28% Higher",
       NULL
} ;

char *gr_strs[] = {      /* Grain menu item strings. */
       "0   Grain Size: 8x8   Gray Levels: 33",
       "1   Grain Size: 8x8   Gray Levels: 33",
       "2   Grain Size: 8x8   Gray Levels: 33",
       "3   Grain Size: 8x8   Gray Levels: 33",
       "4   Grain Size: 6x6   Gray Levels: 37",
       "5   Grain Size: 5x5   Gray Levels: 26",
       "6   Grain Size: 5x5   Gray Levels: 18",
       "7   Grain Size: 4x4   Gray Levels: 17",
       "8   Grain Size: 4x4   Gray Levels: 17",
       "9   Grain Size: 4x4   Gray Levels: 17",
       "10  Grain Size: 3x3   Gray Levels: 10",
       "11  Grain Size: 2x2   Gray Levels: 5",
       NULL
} ;

char *help_strs[] = {   /* Help menu item strings. */
       "Brightness",
       "Contrast",
       "Frame",
       "Grain",
       "Mode",
       "Resolution",
       "Scan",
       "Compress",
       "Port",
       "Speed",
       "Picture",
       NULL
} ;

char *res_strs[] = {     /* Resolution menu item strings. */
       "0   300 dpi  Scale: Full size",
       "1   285 dpi  Scale: 95%",
       "2   270 dpi  Scale: 90%",
       "3   255 dpi  Scale: 85%",
       "4   240 dpi  Scale: 80%",
       "5   225 dpi  Scale: 75%",
       "6   210 dpi  Scale: 70%",
       "7   200 dpi  Scale: 66%",
       "8   180 dpi  Scale: 60%",
       "9   165 dpi  Scale: 55%",
       "10  150 dpi  Scale: 50%",
       "11  135 dpi  Scale: 45%",
       "12  120 dpi  Scale: 40%",
       "13  100 dpi  Scale: 33%",
       "14   90 dpi  Scale: 30%",
       "15   75 dpi  Scale: 25%",
       NULL
} ;

char *set_strs[] = {      /* Set menu item strings. */
       "Picture Name",
       "Default Settings",
       "Clear Scanning Frame",
       "New Scanning Frame",
       "Show Current Settings",
       NULL
} ;

char helpname[MAXLINE] ;      /* Current help file to display. */
char last_message[MAXLINE] ;  /* Last message that was make_displayed. */
char old_picname[MAXLINE] ;   /* Previous picture name. */
char output[MAXLINE] ;        /* Panel_set string to be output. */
char picname[MAXLINE] ;       /* Name of file for raster image. */
char progname[MAXLINE] ;      /* Name of this program. */

int ops[MAXOPS] ;             /* Rasterop functions. */

int switches[4] =
      {
        0,                    /* Mode (Line Art). */
        1,                    /* Data Transfer (Compressed). */
        0,                    /* Serial Port (A). */
        1                     /* Baud Rate (19200). */
      } ;

int framevals[4] =            /* Initial frame in 1/8th inch intervals. */
      {
        16,                   /* X1. */
        16,                   /* Y1. */
        48,                   /* X2. */
        48,                   /* Y2. */
     } ;

int butx ;                    /* Top left X position of popup button. */
int buty ;                    /* Top left Y position of popup button. */
int cur_ch ;                  /* Last keyboard character pressed. */
int curx ;                    /* Current X position of the mouse. */
int cury ;                    /* Current Y position of the mouse. */
int font_width ;              /* Width of a single character. */
int nohelp = 0 ;              /* Indicates if a help file was found. */
int pid ;                     /* Process id of the child scan process. */
int scanning = 0 ;            /* Set if we are in the middle of a scan. */
int showing = 0 ;             /* Set if we are in the middle of a show. */
int type ;                    /* Current pseudo event type. */
int width ;                   /* Width in pixels of scantool window. */

/* Cyclic switches. */
int brightness ;              /* Brightness value. */
int contrast ;                /* Contrast value. */
int grain ;                   /* Grain value. */
int resolution ;              /* Resolution value. */

enum draw_type drawstate ;    /* Current action if a redraw is needed. */

FILE *hfd ;                   /* File descriptor for help file. */


main(argc, argv)
int argc ;
char *argv[] ;
{
  STRCPY(progname, argv[0]) ;      /* Save program name. */
  get_options(argc, argv) ;        /* Extract command line options. */
  init_ws_type() ;                 /* Setup pseudo rasterop functions. */
  drawstate = DO_NOTHING ;         /* No popups currently displayed. */
  if ((hfd = fopen(HELPNAME, "r")) == NULL)
    {
      FPRINTF(stderr, "Help file %s not found\r\n", HELPNAME) ;
      nohelp = 1 ;
    }

  STRCPY(picname, "/home1/falcon/gunnar/scan.rf") ;
  set_defaults() ;                /* Set scanner settings to default. */
  make_frames(argc, argv) ;  /* Create main scantool frame and show window. */
  make_subframes() ;         /* Create canvases for each window. */
  init_fonts() ;             /* Open fonts used by scantool. */
  make_menus() ;             /* Create scantool menubar pulldown menus. */
  start_tool() ;             /* Fit window and wait for events. */
}
